README.md
# aiChat

I used openAI's API KEY so first you should configure your API KEY

# How to install
npm install

# How to run
npm run start:dev


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.3.1](https://github.com/actuallyzefe/aiChat/compare/v1.3.0...v1.3.1) (2023-03-29)

## [1.3.0](https://github.com/actuallyzefe/aiChat/compare/v1.2.3...v1.3.0) (2023-03-29)


### Features

* added TRY CATCH block ([d2cae7e](https://github.com/actuallyzefe/aiChat/commit/d2cae7e099713b1c3cea05848598b6297bf8d615))

### [1.2.3](https://github.com/actuallyzefe/aiChat/compare/v1.2.2...v1.2.3) (2023-03-29)

### [1.2.2](https://github.com/actuallyzefe/aiChat/compare/v1.2.1...v1.2.2) (2023-03-29)

### [1.2.1](https://github.com/actuallyzefe/aiChat/compare/v1.2.0...v1.2.1) (2023-03-25)

## [1.2.0](https://github.com/actuallyzefe/aiChat/compare/v1.1.1...v1.2.0) (2023-03-24)


### Features

* added static files ([40aaa6d](https://github.com/actuallyzefe/aiChat/commit/40aaa6d4dd2424dd2e1e84aab559a50b970adf22))

### [1.1.1](https://github.com/actuallyzefe/aiChat/compare/v1.1.0...v1.1.1) (2023-03-24)


### Bug Fixes

* fixed linebreaks from response ([7aa737b](https://github.com/actuallyzefe/aiChat/commit/7aa737b4391df56d89d93e2826e35503b2c25438))

## [1.1.0](https://github.com/actuallyzefe/aiChat/compare/v1.0.4...v1.1.0) (2023-03-23)


### Features

* added chatWithAi function and route. ([d454d53](https://github.com/actuallyzefe/aiChat/commit/d454d53f3ad372e4b0c9c503a538d6288e1a3e55))

### [1.0.4](https://github.com/actuallyzefe/aiChat/compare/v1.0.3...v1.0.4) (2023-03-23)

### [1.0.3](https://github.com/actuallyzefe/aiChat/compare/v1.0.2...v1.0.3) (2023-03-23)

### [1.0.2](https://github.com/actuallyzefe/aiChat/compare/v1.0.1...v1.0.2) (2023-03-23)

### 1.0.1 (2023-03-22)


controllers/aiController.ts
import {
  Configuration,
  CreateChatCompletionResponse,
  CreateChatCompletionResponseChoicesInner,
  CreateCompletionResponseUsage,
  OpenAIApi,
} from 'openai';
import { Request, Response } from 'express';
import {
  AxiosRequestConfig,
  AxiosResponse,
  AxiosResponseHeaders,
  InternalAxiosRequestConfig,
  RawAxiosResponseHeaders,
} from 'axios';

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

export const generateImage = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { prompt, size } = req.body;
  const imageSize =
    size === 'small' ? '256x256' : size === 'medium' ? '512x512' : '1024x1024';
  try {
    const response = await openai.createImage({
      prompt,
      n: 1,
      size: imageSize,
    });

    const imageUrl = response.data.data[0].url;

    res.status(200).json({
      status: 'Success',
      data: imageUrl,
    });
  } catch (error: any) {
    if (error.response) {
      console.log(error.response.status);
      console.log(error.response.data);
    } else {
      console.log(error.message);
    }
    res.status(400).json({
      status: 'Fail',
      error: 'Image could not be generated',
    });
  }
};

interface Res extends AxiosResponse<CreateChatCompletionResponse, any> {
  data: CreateChatCompletionResponse;
  status: number;
  statusText: string;
  headers: RawAxiosResponseHeaders | AxiosResponseHeaders;
  config: any;
  request?: any;
}

export const chat = async (req: Request, res: Response) => {
  const { prompt } = req.body;
  try {
    const response: Res = await openai.createChatCompletion({
      model: 'gpt-3.5-turbo',
      messages: [{ role: 'user', content: prompt }],
    });

    // MANUAL TYPE CHECK
    // let object = response.data.choices[0].message;
    // if (object !== undefined) {
    //   const aiRes = object.content.replace(/(\r\n|\n|\r)/gm, '');
    //   res.status(200).json({
    //     status: 'Success',
    //     response: aiRes,
    //   });
    // }

    // OPTIONAL CHAINING
    const aiRes = response.data.choices[0]?.message?.content?.replace(
      /(\r\n|\n|\r)/gm,
      ''
    );

    if (aiRes === undefined) {
      res.status(400).json({
        status: 'Fail',
        msg: 'An error happened',
      });
    } else {
      res.status(200).json({
        status: 'Success',
        response: aiRes,
      });
    }
  } catch (error: any) {
    if (error.response) {
      console.log(error.response.status);
      console.log(error.response.data);
    } else {
      console.log(error.message);
    }
    res.status(400).json({
      status: 'Fail',
      error: 'An error happened',
    });
  }
};


controllers/aiController.js
const { Configuration, OpenAIApi } = require('openai');
const readline = require('readline');

// Defining AI
const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

// Image Generator Func
exports.generateImage = async (req, res) => {
  const { prompt, size } = req.body;
  const imageSize =
    size === 'small' ? '256x256' : size === 'medium' ? '512x512' : '1024x1024';
  try {
    const response = await openai.createImage({
      prompt,
      n: 1,
      size: imageSize,
    });

    const imageUrl = response.data.data[0].url;

    res.status(200).json({
      status: 'Success',
      data: imageUrl,
    });
  } catch (error) {
    if (error.response) {
      console.log(error.response.status);
      console.log(error.response.data);
    } else {
      console.log(error.message);
    }
    res.status(400).json({
      status: 'Fail',
      error: 'Image could not be generated',
    });
  }
};

// Chat with AI
exports.chat = async (req, res) => {
  const { prompt } = req.body;
  const response = await openai.createChatCompletion({
    model: 'gpt-3.5-turbo',
    messages: [{ role: 'user', content: prompt }],
  });
  const aiRes = response.data.choices[0].message.content.replace(
    /(\r\n|\n|\r)/gm,
    ''
  );

  // res.render('output', { response: aiRes });
  res.status(200).json({
    status: 'Success',
    response: aiRes,
  });
};

// const userInterface = readline.createInterface({
//   input: process.stdin,
//   output: process.stdout,
// });

// userInterface.prompt();
// userInterface.on("line", async (input) => {
//   const response = await openAi.createChatCompletion({
//     model: "gpt-3.5-turbo",
//     messages: [{ role: "user", content: input }],
//   });
//   console.log(response.data.choices[0].message.content);
//   userInterface.prompt();
// });
